src/markdown/generator.ml

```diff
@@ -0,0 +1,382 @@
+open Odoc_document
+open Types
+open Doctree
+
+module Markup : sig
+  type t
+
+  val noop : t
+
+  val break : t
+
+  val nbsp : t
+
+  val space : t
+
+  val backticks : t
+
+  val open_sq_bracket : t
+
+  val close_sq_bracket : t
+
+  val ( ++ ) : t -> t -> t
+
+  val concat : t list -> t
+
+  val inline' : string list -> t
+
+  val block' : t list -> t
+
+  val list : ?sep:t -> t list -> t
+
+  val anchor' : string -> t
+
+  val string : string -> t
+
+  val str : ('a, unit, string, t) format4 -> 'a
+
+  val escaped : ('a, unit, string, t) format4 -> 'a
+
+  val open_parenthesis : t
+
+  val close_parenthesis : t
+
+  val pp : Format.formatter -> t -> unit
+end = struct
+  type t =
+    | Block of t list
+    | Concat of t list
+    | Break
+    | Space
+    | Anchor of string
+    | String of string
+    | Backticks
+    | Nbsp
+    | OpenSqBracket
+    | CloseSqBracket
+    | OpenParenthesis
+    | CloseParenthesis
+
+  let noop = Concat []
+
+  let break = Break
+
+  let nbsp = Nbsp
+
+  let space = Space
+
+  let backticks = Backticks
+
+  let open_sq_bracket, close_sq_bracket = (OpenSqBracket, CloseSqBracket)
+
+  let open_parenthesis, close_parenthesis = (OpenParenthesis, CloseParenthesis)
+
+  let append t1 t2 =
+    match (t1, t2) with
+    | Concat l1, Concat l2 -> Concat (l1 @ l2)
+    | Concat l1, e2 -> Concat (l1 @ [ e2 ])
+    | e1, Concat l2 -> Concat (e1 :: l2)
+    | e1, e2 -> Concat [ e1; e2 ]
+
+  let ( ++ ) = append
+
+  let concat ts = Concat ts
+
+  let rec intersperse ~sep = function
+    | [] -> []
+    | [ h ] -> [ h ]
+    | h1 :: (_ :: _ as t) -> h1 :: sep :: intersperse ~sep t
+
+  let list ?(sep = Concat []) l = concat @@ intersperse ~sep l
+
+  let anchor' s = Anchor s
+
+  let string s = String s
+
+  let block' ts = Block ts
+
+  let inline' l = List.map (fun s -> string s) l |> concat
+
+  let str fmt = Format.ksprintf (fun s -> string s) fmt
+
+  let escaped fmt = Format.ksprintf (fun s -> string s) fmt
+
+  let rec pp fmt t =
+    match t with
+    | Block b ->
+        let inner = function
+          | [] -> ()
+          | [ x ] -> Format.fprintf fmt "%a" pp x
+          | x :: xs -> Format.fprintf fmt "%a@\n%a" pp x pp (Block xs)
+        in
+        inner b
+    | Concat l -> pp_many fmt l
+    | Break -> Format.fprintf fmt "@\n"
+    | Space -> Format.fprintf fmt " "
+    | Anchor s -> Format.fprintf fmt "<a id=\"%s\"></a>" s
+    | String s -> Format.fprintf fmt "%s" s
+    (* We use double backticks to take care of polymorphic variants or content
+       within backtick, and the spaces before and after the backticks for
+       clarity on what should be enclosed in backticks. For example,
+       "type nums = [ | `One | `Two ]" would be rendered as "``|`````Monday`` "
+       if the spaces were missing.
+    *)
+    | Backticks -> Format.fprintf fmt " `` "
```

Comment:
Should use a single backtick when possible (when there is no backticks in the enclosed code).